### PR TITLE
feat: implement create region request for metric engine

### PR DESCRIPTION
**Pull request by waynexia**
I hereby agree to the terms of the [GreptimeDB CLA](https://gist.github.com/xtang/6378857777706e568c1949c7578592cc)

## What's changed and what's your intention?

- Implement basic logic (add column / add table / get) for metadata region
- Implement create request handler for metric engine 

## Checklist

- [x]  I have written the necessary rustdoc comments.
- [x]  I have added the necessary unit tests and integration tests.

## Refer to a related PR or issue link (optional)

#2070

### CODE REVIEW

File: src/metric-engine/src/engine.rs [hunk 1]
```diff
@@ -41,12 +42,20 @@ pub const METRIC_DATA_REGION_GROUP: RegionGroup = 0;
 /// region group value for metadata region inside a metric region
 pub const METRIC_METADATA_REGION_GROUP: RegionGroup = 1;
 
-const METADATA_SCHEMA_TIMESTAMP_COLUMN_NAME: &str = "ts";
-const METADATA_SCHEMA_KEY_COLUMN_NAME: &str = "k";
-const METADATA_SCHEMA_VALUE_COLUMN_NAME: &str = "val";
+pub const METADATA_SCHEMA_TIMESTAMP_COLUMN_NAME: &str = "ts";
+pub const METADATA_SCHEMA_KEY_COLUMN_NAME: &str = "k";
+pub const METADATA_SCHEMA_VALUE_COLUMN_NAME: &str = "val";
```
Comment (killme2008) [inline src/metric-engine/src/engine.rs:47]: What about `v`? `key` is shortened to `k`.

File: src/metric-engine/src/engine.rs [hunk 2]
```diff
@@ -41,12 +42,20 @@ pub const METRIC_DATA_REGION_GROUP: RegionGroup = 0;
...
 
-const METADATA_REGION_SUBDIR: &str = "metadata";
-const DATA_REGION_SUBDIR: &str = "data";
+pub const METADATA_SCHEMA_TIMESTAMP_COLUMN_INDEX: usize = 0;
+pub const METADATA_SCHEMA_KEY_COLUMN_INDEX: usize = 1;
+pub const METADATA_SCHEMA_VALUE_COLUMN_INDEX: usize = 2;
+
+/// Column name of internal column `__metric_name` that stores the original metric name
+pub const DATA_SCHEMA_METRIC_NAME_COLUMN_NAME: &str = "__metric_name";
```
Comment (killme2008) [inline src/metric-engine/src/engine.rs:54]: Looks better
```suggestion
pub const DATA_SCHEMA_METRIC_NAME_COLUMN_NAME: &str = "__metric";
```

File: src/metric-engine/src/engine.rs [hunk 3]
```diff
@@ -239,13 +275,18 @@ impl MetricEngineInner {
...
             options: HashMap::new(),
             region_dir: metadata_region_dir,
         }
     }
 
-    // todo: register "tag columns" to metadata
+    /// Convert [RegionCreateRequest] for data region.
+    ///
+    /// All tag columns in the original request will be converted to value columns.
+    /// Those columns real semantic type is stored in metadat region.
```
Comment (killme2008) [inline src/metric-engine/src/engine.rs:287]:
```suggestion
    /// Those columns real semantic type is stored in metadata region.
```

File: src/metric-engine/src/metadata_region.rs [hunk 4]
```diff
@@ -60,23 +77,46 @@ impl MetadataRegion {
 
     /// Add a new column key to metadata.
     ///
-    /// This method won't check if the column already exists.
-    pub fn add_column(
+    /// This method won't check if the column already exists. But
+    /// will return if the column is successfully added.
+    pub async fn add_column(
```
Comment (killme2008) [inline src/metric-engine/src/metadata_region.rs:82]:
```suggestion
    #[must_use]
    pub async fn add_column(
```

File: src/metric-engine/src/metadata_region.rs [hunk 5]
```diff
@@ -136,24 +176,139 @@ impl MetadataRegion {
 impl MetadataRegion {
     /// Put if not exist, return if this put operation is successful (error other
     /// than "key already exist" will be wrapped in [Err]).
-    pub fn put_conditionally(
+    pub async fn put_conditionally(
```
Comment (killme2008) [inline src/metric-engine/src/metadata_region.rs:179]:
```suggestion
   #[must_use]
    pub async fn put_conditionally(
```

File: src/metric-engine/src/metadata_region.rs [hunk 6]
```diff
@@ -136,24 +176,139 @@ impl MetadataRegion {
...
+        let Some(first_batch) = scan_result.first() else {
+            return Ok(None);
+        };
+
+        // safety: only value column is projected in the result batch.
+        let value_col = first_batch
+            .column(0)
+            .as_any()
+            .downcast_ref::<StringVector>()
+            .unwrap();
```
Comment (killme2008) [inline src/metric-engine/src/metadata_region.rs:238]:
```suggestion
```

File: src/metric-engine/src/metadata_region.rs [hunk 7]
```diff
@@ -136,24 +176,139 @@ impl MetadataRegion {
...
+            .context(MitoReadOperationSnafu)?;
+        let scan_result = collect(record_batch_stream)
+            .await
+            .context(CollectRecordBatchStreamSnafu)?;
+
+        let exist = !scan_result.is_empty() && scan_result.first().unwrap().num_rows() != 0;
+        Ok(exist)
+    }
+
+    pub async fn get(&self, region_id: RegionId, key: &str) -> Result<Option<String>> {
```
Comment (killme2008) [inline src/metric-engine/src/metadata_region.rs:218]: Adds a comment for this function

File: src/metric-engine/src/metadata_region.rs (hunk 4 above)
Comment (waynexia) [reply src/metric-engine/src/metadata_region.rs:82]: `Result<T>` is marked as `#[must_use]` by default. Do you want to add this annotation to the internal `bool`? Seems only a global clippy warning can do that, which was previously enabled and removed shortly.

**Review by killme2008 — APPROVED** [review]
LGTM

**Review by fengjiachun — APPROVED** [review]
LGTM

**Review by fengjiachun — APPROVED** [review]
LGTM

**Pull request merged**